R/bundle.R

```diff
@@ -933,50 +937,58 @@ snapshotRDependencies <- function(appDir, implicit_dependencies = c(), verbose =
     name = names(named.repos),
     url = as.character(named.repos),
     contrib.url = contrib.url(named.repos, type = "source"),
-    stringsAsFactors = FALSE)
-
-  # get packages records defined in the lockfile
-  records <- utils::tail(df, -1)
+    stringsAsFactors = FALSE
+  )
 
-  # if the package is in a named CRAN-like repository capture it
+  # Sources are created by packrat:
+  # https://github.com/rstudio/packrat/blob/main/R/pkg.R#L328
   tmp <- lapply(seq_len(nrow(records)), function(i) {
-
     pkg <- records[i, "Package"]
     source <- records[i, "Source"]
     repository <- NA
-    # capture Bioconcutor repository
-    if (identical(source, "Bioconductor")) {
-      if (pkg %in% biocPackages) {
-        repository <- biocPackages[pkg, "Repository"]
-      }
-    } else if (isSCMSource(source)) {
-      # leave source+SCM packages alone.
+    if (identical(source, "Bioconductor") && pkg %in% biocPackages) {
+      # installed from known Bioconductor repo
+      repository <- biocPackages[pkg, "Repository"]
+    } else if (source %in% c("github", "bitbucket", "gitlab")) {
+      # leave SCM packages alone.
     } else if (pkg %in% rownames(repo.packages)) {
       # capture CRAN-like repository
 
-      # Find this package in the set of available packages then use its
-      # contrib.url to map back to the configured repositories.
-      package.contrib <- repo.packages[pkg, "Repository"]
-      package.repo.index <- vapply(repo.lookup$contrib.url,
-                                   function(url) grepl(url, package.contrib, fixed = TRUE), logical(1))
-      package.repo <- repo.lookup[package.repo.index, ][1, ]
-      # If the incoming package comes from CRAN, keep the CRAN name in place
-      # even if that means using a different name than the repos list.
-      #
-      # The "cran" source is a well-known location for shinyapps.io.
-      #
-      # shinyapps.io isn't going to use the manifest-provided CRAN URL,
-      # but other consumers (Connect) will.
-      if (tolower(source) != "cran") {
-        source <- package.repo$name
+      # If the name of a source-installed package matches a CRAN package,
+      # packrat automatically fills in the repository information. But that
+      # causes problems if you're working with a developement version that's
+      # newer than CRAN.
+      repo_version <- package_version(repo.packages[pkg, "Version"])
```

Comment:
What defines `package_version`? Could you explain how this will cope with archived packages (e.g. `Kmisc`, from https://github.com/rstudio/rsconnect/issues/508)?

Reply:
`package_version()` is the base R function that takes a string and turns it into a version object so that `1.2.3` compares correctly to `1.10.4`.

Good question about archived packages. I will investigate.

Reply:
Thanks (on both). I'm more used to seeing `packageVersion` and forget about `package_version`.

Reply:
Ok, I think I've fixed it but I need to figure out how to test it and that feels like a post-weekend problem 😄